hibernate-core/src/main/java/org/hibernate/engine/internal/StatefulPersistenceContext.java

```diff
@@ -631,7 +654,13 @@ private void reassociateProxy(LazyInitializer li, HibernateProxy proxy) {
 					.getEntityDescriptor( li.getEntityName() );
 			final EntityKey key = session.generateEntityKey( li.getInternalIdentifier(), persister );
 		  	// any earlier proxy takes precedence
-			getOrInitializeProxiesByKey().putIfAbsent( key, proxy );
+			final EntityHolderImpl holder = getOrInitializeEntitiesByKey().putIfAbsent(
+					key,
+					EntityHolderImpl.forProxy( proxy )
```

Comment:
This is likely not very hot so you can probably ignore me - but if I'm wrong we might want to avoid allocating a new holder object if we don't know if it will be used?

Reply:
We can surely try to improve this further, but I think that a custom map approach that I suggested below would ultimately solve this as well.